dll/win32/shell32/folders/CDesktopFolder.cpp

```diff
@@ -24,6 +24,223 @@
+            }
+            ILFree(pidlLocation);
+        }
+        else
+        {
+            if (attrs)
+                hr = SHGetNameAndFlagsW(pidlLocation, 0, NULL, 0, &attrs);
+
+            if (FAILED_UNEXPECTEDLY(hr))
+            {
+                ILFree(pidlLocation);
+            }
+            else
+            {
+                if (pchEaten)
+                    *pchEaten = cch;
+                *ppidl = pidlLocation;
+            }
+        }
+    }
+
+    // FIXME: SHWindowsPolicy
+    if (SUCCEEDED(hr) && (attrs & SFGAO_STREAM) &&
+        !BindCtx_ContainsObject(pbc, L"Parse Shell Protocol To File Objects"))
+    {
+        ILFree(*ppidl);
+        *ppidl = NULL;
+        hr = HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND);
+    }
+
+    if (pdwAttributes)
+        *pdwAttributes = attrs;
+
+    // FIXME: SHWindowsPolicy
+    if (FAILED(hr) && !Shell_FailForceReturn(hr))
+        hr = HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND);
+
+    return hr;
+}
+
+STDMETHODIMP
+CFileUrlStub::ParseDisplayName(HWND hwndOwner, LPBC pbc, LPOLESTR lpszDisplayName, DWORD *pchEaten,
+                               PIDLIST_RELATIVE *ppidl, DWORD *pdwAttributes)
+{
+    return E_NOTIMPL;
+}
+
+STDMETHODIMP
+CIDListUrlStub::ParseDisplayName(HWND hwndOwner, LPBC pbc, LPOLESTR lpszDisplayName, DWORD *pchEaten,
+                                 PIDLIST_RELATIVE *ppidl, DWORD *pdwAttributes)
+{
+    return E_NOTIMPL;
+}
+
+STDMETHODIMP
+CHttpUrlStub::ParseDisplayName(HWND hwndOwner, LPBC pbc, LPOLESTR lpszDisplayName, DWORD *pchEaten,
+                               PIDLIST_RELATIVE *ppidl, DWORD *pdwAttributes)
+{
+    return E_NOTIMPL;
+}
+
+BOOL CDesktopFolder::_TryUrlJunctions(
+    LPCWSTR pcszURL,
+    IBindCtx *pBindCtx,
+    IShellFolder **ppShellFolder,
+    LPITEMIDLIST *ppidl)
+{
+    PARSEDURLW ParsedURL = { sizeof(ParsedURL) };
+    ::ParseURLW(pcszURL, &ParsedURL);
+
+    *ppShellFolder = NULL;
+    switch (ParsedURL.nScheme)
+    {
+        case URL_SCHEME_FILE:
+            *ppShellFolder = &m_FileUrlStub;
+            break;
+
+        case URL_SCHEME_HTTP:
+        case URL_SCHEME_HTTPS:
+            if (!BindCtx_ContainsObject(pBindCtx, L"Parse Prefer Folder Browsing"))
+                break;
+            *ppShellFolder = &m_HttpUrlStub;
+            break;
+
+        case URL_SCHEME_SHELL:
+            *ppShellFolder = &m_ShellUrlStub;
+            break;
+
+        case URL_SCHEME_MSSHELLROOTED:
+            *ppShellFolder = NULL; // FIXME
+            break;
+
+        case URL_SCHEME_MSSHELLIDLIST:
+            *ppShellFolder = &m_IDListUrlStub;
+            break;
+
+        default:
+            break;
+    }
+
+    return !!*ppShellFolder;
+}
+
+BOOL CDesktopFolder::_GetParentForParsing(
+    LPCWSTR pszPath,
+    IBindCtx *pbc,
+    IShellFolder **ppParentFolder,
+    LPITEMIDLIST *ppidlParent)
+{
+    WCHAR wch = *pszPath;
+    if (pszPath[1] == ':' && ((L'A' <= wch && wch <= L'Z') || (L'a' <= wch && wch <= L'z')))
```

Comment:
Is it safe to access [1] here first? What if the string is empty?

Reply:
Fixed.